rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -96,150 +105,327 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
+                    // This means that user supplied neither so we return everything.
+                    None => None,
+                }
+            }
+        };
 
-        let record_segment_reader = match RecordSegmentReader::from_segment(
+        // Materialize logs.
+        let record_segment_reader: Option<RecordSegmentReader>;
+        match RecordSegmentReader::from_segment(
             &input.record_segment_definition,
             &input.blockfile_provider,
         )
         .await
         {
-            Ok(reader) => reader,
+            Ok(reader) => {
+                record_segment_reader = Some(reader);
+            }
             Err(e) => {
                 match *e {
+                    // Uninitialized segment is fine and means that the record
+                    // segment is not yet initialized in storage.
                     RecordSegmentReaderCreationError::UninitializedSegment => {
-                        // This means no compaction has occured, so we can just return whats on the log.
-                        return Ok(MergeMetadataResultsOperatorOutput {
-                            ids,
-                            metadata,
-                            documents,
-                        });
+                        record_segment_reader = None;
                     }
-                    RecordSegmentReaderCreationError::BlockfileOpenError(_) => {
-                        error!("Error creating Record Segment: {:?}", e);
+                    RecordSegmentReaderCreationError::BlockfileOpenError(e) => {
+                        tracing::error!("Error creating record segment reader {}", e);
                         return Err(
-                            MergeMetadataResultsOperatorError::RecordSegmentCreationError(*e),
+                            MergeMetadataResultsOperatorError::RecordSegmentCreationError(
+                                RecordSegmentReaderCreationError::BlockfileOpenError(e),
+                            ),
                         );
                     }
                     RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                        error!("Error creating Record Segment: {:?}", e);
+                        tracing::error!("Error creating record segment reader {}", e);
                         return Err(
-                            MergeMetadataResultsOperatorError::RecordSegmentCreationError(*e),
+                            MergeMetadataResultsOperatorError::RecordSegmentCreationError(
+                                RecordSegmentReaderCreationError::InvalidNumberOfFiles,
+                            ),
                         );
                     }
-                }
+                };
+            }
+        };
+        // Step 0.5: Get the current max offset id for materialization.
+        // Offset Ids start from 1.
+        let mut curr_max_offset_id = Arc::new(AtomicU32::new(1));
```

Comment:
this is really icky. Why does a read path need to futz with max offset id?

Reply:
yeah I had a task (ofc in my head and not written down anywhere) to make the materializer itself get the max offset id instead. I will do that now